### PR TITLE
Add ADXRS290 driver

### PR DESCRIPTION
**Pull request by MaxAgneni**
Signed-off-by: Massimiliano Agneni <massimiliano.agneni@iptronix.com>

### CODE REVIEW

File: drivers/sensor/adxrs290/adxrs290.c [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+#include <init.h>
+#include <gpio.h>
+#include <misc/byteorder.h>
+#include <misc/__assert.h>
+#include <spi.h>
+
+#include "adxrs290.h"
+
+const uint8_t READ  = 0x80;     /* 0b1000 0000 read command */
+const uint8_t WRITE = 0x7F;     /* 0b01111111 write command */
```
Comment (bogdan-davidoaia) [inline drivers/sensor/adxrs290/adxrs290.c]: Just use a ADXRS290_WRITE define.

File: drivers/sensor/adxrs290/adxrs290.c [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+ * Read from or write to register from the SCP1000:
+ */
+unsigned int adxrs290_read_byte(uint8_t thisRegister, int bytesToRead)
+{
+	unsigned int result;
+
+	/* now combine the address and the command into one byte */
+	spi_tx_buf[0] = thisRegister | READ;
+
+	spi_slave_select(adxrs290_data.spi, adxrs290_data.spi_slave);
```
Comment (bogdan-davidoaia) [inline drivers/sensor/adxrs290/adxrs290.c]: You should check the return values for these SPI functions.

File: drivers/sensor/adxrs290/adxrs290.c [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+
+int adxrs290_init(struct device *dev)
+{
+	struct adxrs290_data *data = dev->driver_data;
+	struct spi_config   spi_config;
+
+	data->spi = device_get_binding(CONFIG_ADXRS290_SPI_DEV_NAME);
+	if (!data->spi) {
+		SYS_LOG_DBG("spi device not found: %s",
+						CONFIG_ADXRS290_SPI_DEV_NAME);
```
Comment (bogdan-davidoaia) [inline drivers/sensor/adxrs290/adxrs290.c]: Fix alignment (also in multiple places in this file).

File: drivers/sensor/adxrs290/adxrs290.h [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+#define ADXRS290_GYR_X_H          0x09
+
+/* 16 2-complement's bits */
+#define ADXRS290_GYR_Y_L          0x0A
+#define ADXRS290_GYR_Y_H          0x0B
+
+/* 12 2-complement's bits */
+#define ADXRS290_TEMP_L           0x0C  /* 7..0 bits */
+#define ADXRS290_TEMP_H           0x0D  /* 11..8 bits */
+#define ADXRS290_TEMP_SCALE_FACT  10.0  /* 10 LSB /centig deg. 0 means 0 deg.*/
```
Comment (bogdan-davidoaia) [inline drivers/sensor/adxrs290/adxrs290.h]: Defined by not used. Also don't see the point in using floating point.

Maybe also define and use a scale factor constant for gyro.

File: drivers/sensor/adxrs290/adxrs290.c [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+					  struct sensor_value *val)
+{
+	struct adxrs290_data *data = dev->driver_data;
+
+	switch (chan) {
+
+	case SENSOR_CHAN_GYRO_X:
+		/* Angular velocity around the X axis, in radians/s. */
+		val->val1 = (int32_t)data->x / 200;
+		val->val2 = ((int32_t)data->x % 200) * 1000000;
```
Comment (bogdan-davidoaia) [inline drivers/sensor/adxrs290/adxrs290.c]: Since you are dividing by 200, you should multiply by 1000000/200. Same for the next two.

File: drivers/sensor/adxrs290/adxrs290.h [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+/*
+ * Low-Pass Filter Pole Locations
+ *  The data is the Frequency in Hz
+ */
+#define ADXRS_LPF_480_HZ          0x00   /*  480_HZ is Default */
+#define ADXRS_LPF_320_HZ          0x01   /*  320_HZ */
+#define ADXRS_LPF_160_HZ          0x02   /*  160_HZ */
+#define ADXRS_LPF_80_HZ           0x03   /*  80_HZ */
+#define ADXRS_LPF_56_6_HZ         0x04   /*  56.6_HZ */
+#define ADXRS_LPF_40_HZ           0x05   /*  40 HZ */
```
Comment (bogdan-davidoaia) [inline drivers/sensor/adxrs290/adxrs290.h]: 40_HZ in the comment to match the rest.

File: drivers/sensor/adxrs290/adxrs290.h [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+ * 01 Data ready, high until read
+ */
+#define ADXRS290_DATA_READY_INT_MASK    0x03
+
+
+/*Sensors Sensitivity */
+
+/*
+ * Low-Pass Filter Pole Locations
+ *  The data is the Frequency in Hz
```
Comment (bogdan-davidoaia) [inline drivers/sensor/adxrs290/adxrs290.h]: Wording seems a bit strange to me.

File: drivers/sensor/adxrs290/adxrs290.h [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+#define ADXRS290_BANDPASS_FILTER  0x11
+#define ADXRS290_BPF_LPF_MASK     0x07
+#define ADXRS290_BPF_HPF_MASK     0xF0
+#define ADXRS290_BPF_HPF_OFFSET   0x4
+
+#define ADXRS290_DATA_READY_REG   0x12
+
+/* Set this bit to get triggered on data ready via interrupt
+ * Set bit to 01 to gen rata ready interrupt
+ * at the sync/asel pin when data becomes avail
```
Comment (bogdan-davidoaia) [inline drivers/sensor/adxrs290/adxrs290.h]: available

File: drivers/sensor/adxrs290/adxrs290.h [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+/* 1: measurement mode, 0 chip in standby mode */
+#define ADXRS290_BANDPASS_FILTER  0x11
+#define ADXRS290_BPF_LPF_MASK     0x07
+#define ADXRS290_BPF_HPF_MASK     0xF0
+#define ADXRS290_BPF_HPF_OFFSET   0x4
+
+#define ADXRS290_DATA_READY_REG   0x12
+
+/* Set this bit to get triggered on data ready via interrupt
+ * Set bit to 01 to gen rata ready interrupt
```
Comment (bogdan-davidoaia) [inline drivers/sensor/adxrs290/adxrs290.h]: to generate data ready interrupt

File: drivers/sensor/adxrs290/adxrs290.c [hunk 10]
```diff
@@ -0,0 +1,416 @@
...
+	return signed_data;
+}
+
+int16_t adxrs290_standby_read_temperature(void)
+{
+	uint16_t data;
+	uint8_t lsb;
+	uint8_t msb;
+
+	// In standby we can read one byte at the time
```
Comment (bogdan-davidoaia) [inline drivers/sensor/adxrs290/adxrs290.c]: Don't use // comments.

**Review by bogdan-davidoaia — CHANGES_REQUESTED** [review]
Also address issues pointed out by https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/448/consolidateTests

**Review by pfalcon — CHANGES_REQUESTED** [review]
This PR seems to have gone off track, accumulating unrelated commits over time. Please use a separate topic branch for each PR.

File: drivers/sensor/si1153/si1153.c [hunk 11]
```diff
@@ -0,0 +1,663 @@
...
+
+	/* Get the response register contents */
+	response = si1153_reg_read(data, SI115x_REG_RESPONSE0);
+	if (response < 0) {
+		return response;
+	}
+
+	response = response & RSP0_COUNTER_MASK;
+
+	/* Double-check the response register is consistent */
```
Comment (bogdan-davidoaia) [inline drivers/sensor/si1153/si1153.c]: Why is this needed?

File: drivers/sensor/si1153/si1153.c [hunk 12]
```diff
@@ -0,0 +1,663 @@
...
+		count++;
+	}
+
+	return 0;
+}
+
+static int si115x_force(struct si1153_data *data)
+{
+	/* Sends a FORCE command to the Si115x/6x */
+	return _send_cmd(data, 0x11);
```
Comment (bogdan-davidoaia) [inline drivers/sensor/si1153/si1153.c]: Use defines for these hex values.

File: drivers/sensor/si1153/si1153.c [hunk 13]
```diff
@@ -0,0 +1,663 @@
...
+	retval += si115x_param_set(data, PARAM_LED1_A, LED_CURRENT_199);
+	retval += si115x_param_set(data, PARAM_LED2_A, LED_CURRENT_199);
+	retval += si115x_param_set(data, PARAM_LED3_A, LED_CURRENT_310);
+
+	retval += si1153_reg_write(data, SI115x_REG_IRQ_ENABLE, 0x0f);
+
+	return retval;
+}
+
+/**
```
Comment (bogdan-davidoaia) [inline drivers/sensor/si1153/si1153.c]: Remote this empty comment.

File: drivers/sensor/si1153/si1153.c [hunk 14]
```diff
@@ -0,0 +1,663 @@
...
+
+		si115x_handler(data, &samples);
+
+		data->gesture = si115x_gesture_algorithm(&samples);
+
+		data->ch0 = samples.ch0;
+		data->ch1 = samples.ch1;
+		data->ch2 = samples.ch2;
+		data->ch3 = samples.ch3;
+#if 0
```
Comment (bogdan-davidoaia) [inline drivers/sensor/si1153/si1153.c]: Instead of using #if 0/1 you can use a DEBUG define at the top of the file.

File: drivers/sensor/si1153/si1153.c [hunk 15]
```diff
@@ -0,0 +1,663 @@
...
+#define ADCPOST     (ADCPOST_24BIT | ADCPOST_POSTSHIFT_0 | ADCPOST_THRESH_EN_0)
+#define MEASURECFG  (MEASCFG_NO_MEAS | MEASCFG_LED_NOM | MEASCFG_BANK_SEL_A)
+
+static int si115x_init(struct si1153_data *data)
+{
+	int retval;
+
+	retval = si115x_reset(data);
+
+	retval += si115x_param_set(data, PARAM_CH_LIST, 0x0f);
```
Comment (bogdan-davidoaia) [inline drivers/sensor/si1153/si1153.c]: I'd use |= instead of += to avoid overflows that could result retval adding up to 0.

File: drivers/sensor/si1153/si1153.c [hunk 16]
```diff
@@ -0,0 +1,663 @@
...
+#include <sensor.h>
+#include <init.h>
+#include <gpio.h>
+#include <misc/byteorder.h>
+#include <misc/__assert.h>
+#include <i2c.h>
+
+#include "si1153.h"
+
+struct SI115X_SAMPLES {
```
Comment (bogdan-davidoaia) [inline drivers/sensor/si1153/si1153.c]: Use lowercase

File: drivers/sensor/si1153/si1153.c [hunk 17]
```diff
@@ -0,0 +1,663 @@
...
+					  enum sensor_channel chan,
+					  struct sensor_value *val)
+{
+	struct si1153_data *data = dev->driver_data;
+
+	switch (chan) {
+
+	case SENSOR_CHAN_LIGHT:
+	case SENSOR_CHAN_IR:
+		val->val1 = data->ch0;
```
Comment (bogdan-davidoaia) [inline drivers/sensor/si1153/si1153.c]: Is this value actually in lux?

File: drivers/sensor/si1153/si1153.c [hunk 18]
```diff
@@ -0,0 +1,663 @@
...
+		val->val2 = 0;
+		break;
+	case SENSOR_CHAN_PROX:
```
Comment (bogdan-davidoaia) [inline drivers/sensor/si1153/si1153.c]: The Sensor API doesn't take into account multiple channels, so this behavior would be against what an application developer would expect.

File: drivers/sensor/si1153/si1153.c [hunk 19]
```diff
@@ -0,0 +1,663 @@
...
+		val->val1 = data->ch1;
+		val->val2 = 0;
+		val++;
+		val->val1 = data->ch2;
+		val->val2 = 0;
+		val++;
+		val->val1 = data->ch3;
+		val->val2 = 0;
+		break;
+	case SENSOR_CHAN_GESTURE:
```
Comment (bogdan-davidoaia) [inline drivers/sensor/si1153/si1153.c]: Should add a patch for defining this channel. Since the other channels (LIGHT, IR, PROX) may not be in line with the definitions in the API, would it make sense to keep just the gesture channel?

**Review by bogdan-davidoaia — CHANGES_REQUESTED** [review]
Also fix style issues.
